{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "3d4a3ca7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#generate ORF\n",
    "\n",
    "from Bio import Entrez\n",
    "\n",
    "from Bio import SeqIO\n",
    "\n",
    "Entrez.email='francisrobitaille07@gmail.com'\n",
    "\n",
    "#mit codons produce a lot more ORFs than cyt ones. (more start codons, less stop ones)\n",
    "cytstop = ['UAA','UAG','UGA']\n",
    "cytstart = ['AUG']\n",
    "mitstop = ['UAA','UAG'] #add tRNA molecular pattern + ribosome stalling patterns?\n",
    "mitnouvstop = ['AGA','AGG']\n",
    "mitstart = ['AUG','AUU','AUA'] #is there molecular evidence, or could it point to some other initiation mechanism?\n",
    "\n",
    "#replace Z with real AA value \n",
    "mit = {\"UUU\": \"F\", \"UUC\": \"F\", \"UUA\": \"L\", \"UUG\": \"L\",\n",
    "    \"CUU\": \"L\", \"CUC\": \"L\", \"CUA\": \"L\", \"CUG\": \"L\",\n",
    "    \"AUU\": \"I\", \"AUC\": \"I\", \"AUA\": \"I\", \"AUG\": \"M\",\n",
    "    \"GUU\": \"V\", \"GUC\": \"V\", \"GUA\": \"V\", \"GUG\": \"V\",\n",
    "    \"UCU\": \"S\", \"UCC\": \"S\", \"UCA\": \"S\", \"UCG\": \"S\",\n",
    "    \"CCU\": \"P\", \"CCC\": \"P\", \"CCA\": \"P\", \"CCG\": \"P\",\n",
    "    \"ACU\": \"T\", \"ACC\": \"T\", \"ACA\": \"T\", \"ACG\": \"T\",\n",
    "    \"GCU\": \"A\", \"GCC\": \"A\", \"GCA\": \"A\", \"GCG\": \"A\",\n",
    "    \"UAU\": \"Y\", \"UAC\": \"Y\", \"UAA\": \"*\", \"UAG\": \"*\",\n",
    "    \"CAU\": \"H\", \"CAC\": \"H\", \"CAA\": \"Q\", \"CAG\": \"Q\",\n",
    "    \"AAU\": \"N\", \"AAC\": \"N\", \"AAA\": \"K\", \"AAG\": \"K\",\n",
    "    \"GAU\": \"D\", \"GAC\": \"D\", \"GAA\": \"E\", \"GAG\": \"E\",\n",
    "    \"UGU\": \"C\", \"UGC\": \"C\", \"UGA\": \"Z\", \"UGG\": \"W\",\n",
    "    \"CGU\": \"R\", \"CGC\": \"R\", \"CGA\": \"R\", \"CGG\": \"R\",\n",
    "    \"AGU\": \"S\", \"AGC\": \"S\", \"AGA\": \"R\", \"AGG\": \"R\",\n",
    "    \"GGU\": \"G\", \"GGC\": \"G\", \"GGA\": \"G\", \"GGG\": \"G\"}\n",
    "\n",
    "mit_nouv = {\"UUU\": \"F\", \"UUC\": \"F\", \"UUA\": \"L\", \"UUG\": \"L\",\n",
    "    \"CUU\": \"L\", \"CUC\": \"L\", \"CUA\": \"L\", \"CUG\": \"L\",\n",
    "    \"AUU\": \"I\", \"AUC\": \"I\", \"AUA\": \"I\", \"AUG\": \"M\",\n",
    "    \"GUU\": \"V\", \"GUC\": \"V\", \"GUA\": \"V\", \"GUG\": \"V\",\n",
    "    \"UCU\": \"S\", \"UCC\": \"S\", \"UCA\": \"S\", \"UCG\": \"S\",\n",
    "    \"CCU\": \"P\", \"CCC\": \"P\", \"CCA\": \"P\", \"CCG\": \"P\",\n",
    "    \"ACU\": \"T\", \"ACC\": \"T\", \"ACA\": \"T\", \"ACG\": \"T\",\n",
    "    \"GCU\": \"A\", \"GCC\": \"A\", \"GCA\": \"A\", \"GCG\": \"A\",\n",
    "    \"UAU\": \"Y\", \"UAC\": \"Y\", \"UAA\": \"Z\", \"UAG\": \"Z\",\n",
    "    \"CAU\": \"H\", \"CAC\": \"H\", \"CAA\": \"Q\", \"CAG\": \"Q\",\n",
    "    \"AAU\": \"N\", \"AAC\": \"N\", \"AAA\": \"K\", \"AAG\": \"K\",\n",
    "    \"GAU\": \"D\", \"GAC\": \"D\", \"GAA\": \"E\", \"GAG\": \"E\",\n",
    "    \"UGU\": \"C\", \"UGC\": \"C\", \"UGA\": \"Z\", \"UGG\": \"W\",\n",
    "    \"CGU\": \"R\", \"CGC\": \"R\", \"CGA\": \"R\", \"CGG\": \"R\",\n",
    "    \"AGU\": \"S\", \"AGC\": \"S\", \"AGA\": \"*\", \"AGG\": \"*\",\n",
    "    \"GGU\": \"G\", \"GGC\": \"G\", \"GGA\": \"G\", \"GGG\": \"G\"}\n",
    "\n",
    "cyt = {\"UUU\": \"F\", \"UUC\": \"F\", \"UUA\": \"L\", \"UUG\": \"L\",\n",
    "    \"CUU\": \"L\", \"CUC\": \"L\", \"CUA\": \"L\", \"CUG\": \"L\",\n",
    "    \"AUU\": \"I\", \"AUC\": \"I\", \"AUA\": \"I\", \"AUG\": \"M\",\n",
    "    \"GUU\": \"V\", \"GUC\": \"V\", \"GUA\": \"V\", \"GUG\": \"V\",\n",
    "    \"UCU\": \"S\", \"UCC\": \"S\", \"UCA\": \"S\", \"UCG\": \"S\",\n",
    "    \"CCU\": \"P\", \"CCC\": \"P\", \"CCA\": \"P\", \"CCG\": \"P\",\n",
    "    \"ACU\": \"T\", \"ACC\": \"T\", \"ACA\": \"T\", \"ACG\": \"T\",\n",
    "    \"GCU\": \"A\", \"GCC\": \"A\", \"GCA\": \"A\", \"GCG\": \"A\",\n",
    "    \"UAU\": \"Y\", \"UAC\": \"Y\", \"UAA\": \"*\", \"UAG\": \"*\",\n",
    "    \"CAU\": \"H\", \"CAC\": \"H\", \"CAA\": \"Q\", \"CAG\": \"Q\",\n",
    "    \"AAU\": \"N\", \"AAC\": \"N\", \"AAA\": \"K\", \"AAG\": \"K\",\n",
    "    \"GAU\": \"D\", \"GAC\": \"D\", \"GAA\": \"E\", \"GAG\": \"E\",\n",
    "    \"UGU\": \"C\", \"UGC\": \"C\", \"UGA\": \"*\", \"UGG\": \"W\",\n",
    "    \"CGU\": \"R\", \"CGC\": \"R\", \"CGA\": \"R\", \"CGG\": \"R\",\n",
    "    \"AGU\": \"S\", \"AGC\": \"S\", \"AGA\": \"R\", \"AGG\": \"R\",\n",
    "    \"GGU\": \"G\", \"GGC\": \"G\", \"GGA\": \"G\", \"GGG\": \"G\"}\n",
    "    \n",
    "\n",
    "#defining functions\n",
    "\n",
    "#returns the complementary strand in the correct order for translation (5' -> 3')\n",
    "def reverse_complement(sequence):\n",
    "    # Define a dictionary for complementing nucleotides\n",
    "    complement = {'A': 'U', 'T': 'A', 'C': 'G', 'G': 'C','N':'N'}\n",
    "    # Reverse the sequence\n",
    "    reversed_sequence = sequence[::-1]\n",
    "    # Complement each nucleotide and join them to form the reverse complement\n",
    "    reverse_complement = ''.join([complement[nucleotide] for nucleotide in reversed_sequence])\n",
    "    return reverse_complement\n",
    "\n",
    "#DOUBLE TRIPLE CHECK MAKE SURE IT WORKS\n",
    "def generate_ORFs(seq, start_codons, stop_codons, filter_cod):\n",
    "    \"Cette fonction fait chier.\"\n",
    "    total_ORF = []\n",
    "    for i in range(len(seq)-3):\n",
    "        codon = seq[i:i+3]\n",
    "        if codon in start_codons:\n",
    "            #initiate ORF\n",
    "            ORF = ''\n",
    "            for nuc in range(i,len(seq),3):\n",
    "                codon = seq[nuc:nuc+3]\n",
    "                if codon not in stop_codons:\n",
    "                    ORF += codon\n",
    "                else:\n",
    "                    ORF += codon\n",
    "                    #minimal altORF length\n",
    "                    if len(ORF)> filter_cod:\n",
    "                        total_ORF.append(ORF)\n",
    "                    break\n",
    "    return total_ORF\n",
    "              \n",
    "def transcribe(dna_seq):\n",
    "    \n",
    "    rna_seq = dna_seq.replace(\"T\", \"U\")\n",
    "    return rna_seq\n",
    "        \n",
    "\n",
    "#need another translate function for mitochondrial (maybe have the codon table as input)\n",
    "def translate(sequence, start_codons, stop_codons, codon_table):\n",
    "   \n",
    "    #make new code in which start and stop codons are derived from the codon table\n",
    "    protein_sequence = ''\n",
    "    i = 0\n",
    "    while i < len(sequence):\n",
    "        codon = sequence[i:i+3]\n",
    "        #if codon_table[codon] == 'M'\n",
    "        if codon in start_codons:\n",
    "            protein_sequence += 'M'\n",
    "        \n",
    "        elif codon in stop_codons:\n",
    "            protein_sequence += '*'\n",
    "        elif codon in codon_table:\n",
    "            protein_sequence += codon_table[codon]\n",
    "        i += 3\n",
    "    return protein_sequence\n",
    "\n",
    "def translate_orfs(orfs,start,stop,codon_table):\n",
    "    prot_orfs = []\n",
    "    for i in range(len(orfs)):\n",
    "        prot_orfs.append(translate(orfs[i],start,stop,codon_table))\n",
    "    return prot_orfs\n",
    "\n",
    "#turn a list of string into a list of seq objects to allow alignement\n",
    "def seq_list(prots_list):\n",
    "    newlist = []\n",
    "    for seq in prots_list:\n",
    "        newlist.append(Seq(seq))\n",
    "        \n",
    "    return newlist\n",
    "\n",
    "# Define a list of protein sequences to compare against\n",
    "\n",
    "def protein_aligner (query, list_of_prots):\n",
    "    \n",
    "    # Initialize a PairwiseAligner object\n",
    "    aligner = PairwiseAligner()\n",
    "\n",
    "    # Loop through the list of sequences to find the best match\n",
    "    best_score = 0\n",
    "    best_seq = None\n",
    "    \n",
    "    for seq in list_of_prots:\n",
    "    # Use the aligner object to get the optimal global alignment between the two sequences\n",
    "        alignments = aligner.align(query, seq)\n",
    "    # Get the alignment score of the first alignment\n",
    "        alignment_score = alignments[0].score\n",
    "    # If the current alignment score is better than the previous best score, update the best score and best sequence\n",
    "        if alignment_score > best_score:\n",
    "            best_score = alignment_score\n",
    "            best_seq = seq\n",
    "    \n",
    "    return best_score\n",
    "\n",
    "#print('best match:   ',best_seq)\n",
    "#print('query peptide:',query)\n",
    "#print('match score:', best_score)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2cfcb7c0",
   "metadata": {},
   "source": [
    "# Test\n",
    "\n",
    "Ceci est du texte entre le code : "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33836141",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "1018d238",
   "metadata": {},
   "outputs": [],
   "source": [
    "#generate sequences\n",
    "# if I enforce tRNA punctuation, I need to split the sequences up here\n",
    "\n",
    "handle = Entrez.efetch(db='nucleotide',id='NC_012920.1', rettype='fasta')\n",
    "record = SeqIO.read(handle, 'fasta')\n",
    "mtdna_GB = str(record.seq)\n",
    "\n",
    "#in the order in which it will be read (5' -> 3'). GB seq is the coding strand (for H strand) so we just switch T -> U.\n",
    "\n",
    "RNA_L = reverse_complement(mtdna_GB)\n",
    "RNA_H = transcribe(mtdna_GB)\n",
    "#print(len(mtdna_L))\n",
    "#print(mtdna_H)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "db3d6ade",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MC H-Strand ORFs:  674\n",
      "MC L-Strand ORFs:  808\n",
      "UGC H-Strand ORFs:  106\n",
      "UGC L-Strand ORFs:  269\n"
     ]
    }
   ],
   "source": [
    "min_orf_length = 30\n",
    "\n",
    "#generate a list of ORFs at the mRNA level\n",
    "orfs_H_mit = generate_ORFs(RNA_H, mitstart, mitstop, min_orf_length)\n",
    "orfs_L_mit = generate_ORFs(RNA_L, mitstart, mitstop, min_orf_length)\n",
    "orfs_H_mit_nouv = generate_ORFs(RNA_H, mitstart, mitnouvstop, min_orf_length)\n",
    "orfs_L_mit_nouv = generate_ORFs(RNA_L, mitstart, mitnouvstop, min_orf_length)\n",
    "orfs_H_cyt = generate_ORFs(RNA_H, cytstart, cytstop, min_orf_length)\n",
    "orfs_L_cyt = generate_ORFs(RNA_L, cytstart, cytstop, min_orf_length)\n",
    "\n",
    "#translate those ORFs into a list of corresponding proteins\n",
    "prots_H = translate_orfs(orfs_H_mit, mitstart,mitstop,mit)\n",
    "prots_L = translate_orfs(orfs_L_mit, mitstart,mitstop,mit)\n",
    "prots_H_nouv = translate_orfs(orfs_H_mit, mitstart,mitnouvstop,mit_nouv)\n",
    "prots_L_nouv = translate_orfs(orfs_L_mit, mitstart,mitnouvstop,mit_nouv)\n",
    "prots_H_cyt = translate_orfs(orfs_H_cyt, cytstart, cytstop, cyt)\n",
    "prots_L_cyt = translate_orfs(orfs_L_cyt, cytstart, cytstop, cyt)\n",
    "\n",
    "\n",
    "print('MC H-Strand ORFs: ',len(orfs_H_mit))\n",
    "print('MC L-Strand ORFs: ',len(orfs_L_mit))\n",
    "print('UGC H-Strand ORFs: ',len(orfs_H_cyt))\n",
    "print('UGC L-Strand ORFs: ',len(orfs_L_cyt))\n",
    "\n",
    "#print('MC H-Strand prots: ', prots_H)\n",
    "#print('MC L-Strand prots: ', prots_L)\n",
    "#print('UGC H-Strand prots: ',prots_H_cyt)\n",
    "#print('UGC L-Strand prots: ',prots_L_cyt)\n",
    "\n",
    "seq_prots_H = seq_list(prots_H)\n",
    "seq_prots_L = seq_list(prots_L)\n",
    "seq_prots_H_nouv = seq_list(prots_H_nouv)\n",
    "seq_prots_L_nouv = seq_list(prots_L_nouv)\n",
    "seq_prots_H_cyt = seq_list(prots_H_cyt)\n",
    "seq_prots_L_cyt = seq_list(prots_H_cyt)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "274ad643",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The refprot CO1 is on the Heavy strand\n",
      "H-Strand: all ORFS and CO1 . best score is: 485.0\n",
      "L-Strand: all ORFS and CO1 . best score is: 132.0\n",
      "The refprot CO2 is on the Heavy strand\n",
      "H-Strand: all ORFS and CO2 . best score is: 217.0\n",
      "L-Strand: all ORFS and CO2 . best score is: 85.0\n",
      "The refprot CO3 is on the Heavy strand\n",
      "H-Strand: all ORFS and CO3 . best score is: 245.0\n",
      "L-Strand: all ORFS and CO3 . best score is: 88.0\n",
      "The refprot ATP6 is on the Heavy strand\n",
      "H-Strand: all ORFS and ATP6 . best score is: 210.0\n",
      "L-Strand: all ORFS and ATP6 . best score is: 90.0\n",
      "The refprot ATP8 is on the Heavy strand\n",
      "H-Strand: all ORFS and ATP8 . best score is: 65.0\n",
      "L-Strand: all ORFS and ATP8 . best score is: 41.0\n",
      "The refprot CYTB is on the Heavy strand\n",
      "H-Strand: all ORFS and CYTB . best score is: 359.0\n",
      "L-Strand: all ORFS and CYTB . best score is: 112.0\n",
      "The refprot ND1 is on the Heavy strand\n",
      "H-Strand: all ORFS and ND1 . best score is: 299.0\n",
      "L-Strand: all ORFS and ND1 . best score is: 102.0\n",
      "The refprot ND2 is on the Heavy strand\n",
      "H-Strand: all ORFS and ND2 . best score is: 330.0\n",
      "L-Strand: all ORFS and ND2 . best score is: 105.0\n",
      "The refprot ND3 is on the Heavy strand\n",
      "H-Strand: all ORFS and ND3 . best score is: 108.0\n",
      "L-Strand: all ORFS and ND3 . best score is: 60.0\n",
      "The refprot ND4 is on the Heavy strand\n",
      "H-Strand: all ORFS and ND4 . best score is: 431.0\n",
      "L-Strand: all ORFS and ND4 . best score is: 123.0\n",
      "The refprot ND4L is on the Heavy strand\n",
      "H-Strand: all ORFS and ND4L . best score is: 95.0\n",
      "L-Strand: all ORFS and ND4L . best score is: 58.0\n",
      "The refprot ND5 is on the Heavy strand\n",
      "H-Strand: all ORFS and ND5 . best score is: 574.0\n",
      "L-Strand: all ORFS and ND5 . best score is: 145.0\n",
      "The refprot ND6 is on the Light strand\n",
      "H-Strand: all ORFS and ND6 . best score is: 110.0\n",
      "L-Strand: all ORFS and ND6 . best score is: 159.0\n"
     ]
    }
   ],
   "source": [
    "from Bio.Align import PairwiseAligner\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "# Define the protein sequence you want to match\n",
    "#known altorfs also\n",
    "\n",
    "CO1 = Seq('MFADRWLFSTNHKDIGTLYLLFGAWAGVLGTALSLLIRAELGQPGNLLGNDHIYNVIVTAHAFVMIFFMVMPIMIGGFGNWLVPLMIGAPDMAFPRMNNMSFWLLPPSLLLLLASAMVEAGAGTGWTVYPPLAGNYSHPGASVDLTIFSLHLAGVSSILGAINFITTIINMKPPAMTQYQTPLFVWSVLITAVLLLLSLPVLAAGITMLLTDRNLNTTFFDPAGGGDPILYQHLFWFFGHPEVYILILPGFGMISHIVTYYSGKKEPFGYMGMVWAMMSIGFLGFIVWAHHMFTVGMDVDTRAYFTSATMIIAIPTGVKVFSWLATLHGSNMKWSAAVLWALGFIFLFTVGGLTGIVLANSSLDIVLHDTYYVVAHFHYVLSMGAVFAIMGGFIHWFPLFSGYTLDQTYAKIHFTIMFIGVNLTFFPQHFLGLSGMPRRYSDYPDAYTTWNILSSVGSFISLTAVMLMIFMIWEAFASKRKVLMVEEPSMNLEWLYGCPPPYHTFEEPVYMKS')\n",
    "CO2 = Seq('MAHAAQVGLQDATSPIMEELITFHDHALMIIFLICFLVLYALFLTLTTKLTNTNISDAQEMETVWTILPAIILVLIALPSLRILYMTDEVNDPSLTIKSIGHQWYWTYEYTDYGGLIFNSYMLPPLFLEPGDLRLLDVDNRVVLPIEAPIRMMITSQDVLHSWAVPTLGLKTDAIPGRLNQTTFTATRPGVYYGQCSEICGANHSFMPIVLELIPLKIFEMGPVFTL')\n",
    "CO3 = Seq('MTHQSHAYHMVKPSPWPLTGALSALLMTSGLAMWFHFHSMTLLMLGLLTNTLTMYQWWRDVTRESTYQGHHTPPVQKGLRYGMILFITSEVFFFAGFFWAFYHSSLAPTPQLGGHWPPTGITPLNPLEVPLLNTSVLLASGVSITWAHHSLMENNRNQMIQALLITILLGLYFTLLQASEYFESPFTISDGIYGSTFFVATGFHGLHVIIGSTFLTICFIRQLMFHFTSKHHFGFEAAAWYWHFVDVVWLFLYVSIYWWGS')\n",
    "ATP6 = Seq('MNENLFASFIAPTILGLPAAVLIILFPPLLIPTSKYLINNRLITTQQWLIKLTSKQMMTMHNTKGRTWSLMLVSLIIFIATTNLLGLLPHSFTPTTQLSMNLAMAIPLWAGTVIMGFRSKIKNALAHFLPQGTPTPLIPMLVIIETISLLIQPMALAVRLTANITAGHLLMHLIGSATLAMSTINLPSTLIIFTILILLTILEIAVALIQAYVFTLLVSLYLHDNT')\n",
    "ATP8 = Seq('MPQLNTTVWPTMITPMLLTLFLITQLKMLNTNYHLPPSPKPMKMKNYNKPWEPKWTKICSLHSLPPQS')\n",
    "CYTB = Seq('MTPMRKTNPLMKLINHSFIDLPTPSNISAWWNFGSLLGACLILQITTGLFLAMHYSPDASTAFSSIAHITRDVNYGWIIRYLHANGASMFFICLFLHIGRGLYYGSFLYSETWNIGIILLLATMATAFMGYVLPWGQMSFWGATVITNLLSAIPYIGTDLVQWIWGGYSVDSPTLTRFFTFHFILPFIIAALATLHLLFLHETGSNNPLGITSHSDKITFHPYYTIKDALGLLLFLLSLMTLTLFSPDLLGDPDNYTLANPLNTPPHIKPEWYFLFAYTILRSVPNKLGGVLALLLSILILAMIPILHMSKQQSMMFRPLSQSLYWLLAADLLILTWIGGQPVSYPFTIIGQVASVLYFTTILILMPTISLIENKMLKWA')\n",
    "ND1 = Seq('MPMANLLLLIVPILIAMAFLMLTERKILGYMQLRKGPNVVGPYGLLQPFADAMKLFTKEPLKPATSTITLYITAPTLALTIALLLWTPLPMPNPLVNLNLGLLFILATSSLAVYSILWSGWASNSNYALIGALRAVAQTISYEVTLAIILLSTLLMSGSFNLSTLITTQEHLWLLLPSWPLAMMWFISTLAETNRTPFDLAEGESELVSGFNIEYAAGPFALFFMAEYTNIIMMNTLTTTIFLGTTYDALSPELYTTYFVTKTLLLTSLFLWIRTAYPRFRYDQLMHLLWKNFLPLTLALLMWYVSMPITISSIPPQT')\n",
    "ND2 = Seq('MNPLAQPVIYSTIFAGTLITALSSHWFFTWVGLEMNMLAFIPVLTKKMNPRSTEAAIKYFLTQATASMILLMAILFNNMLSGQWTMTNTTNQYSSLMIMMAMAMKLGMAPFHFWVPEVTQGTPLTSGLLLLTWQKLAPISIMYQISPSLNVSLLLTLSILSIMAGSWGGLNQTQLRKILAYSSITHMGWMMAVLPYNPNMTILNLTIYIILTTTAFLLLNLNSSTTTLLLSRTWNKLTWLTPLIPSTLLSLGGLPPLTGFLPKWAIIEEFTKNNSLIIPTIMATITLLNLYFYLRLIYSTSITLLPMSNNVKMKWQFEHTKPTPFLPTLIALTTLLLPISPFMLMIL')\n",
    "ND3 = Seq('MNFALILMINTLLALLLMIITFWLPQLNGYMEKSTPYECGFDPMSPARVPFSMKFFLVAITFLLFDLEIALLLPLPWALQTTNLPLMVMSSLLLIIILALSLAYEWLQKGLDWTE')\n",
    "ND4 = Seq(\"MLKLIVPTIMLLPLTWLSKKHMIWINTTTHSLIISIIPLLFFNQINNNLFSCSPTFSSDPLTTPLLMLTTWLLPLTIMASQRHLSSEPLSRKKLYLSMLISLQISLIMTFTATELIMFYIFFETTLIPTLAIITRWGNQPERLNAGTYFLFYTLVGSLPLLIALIYTHNTLGSLNILLLTLTAQELSNSWANNLMWLAYTMAFMVKMPLYGLHLWLPKAHVEAPIAGSMVLAAVLLKLGGYGMMRLTLILNPLTKHMAYPFLVLSLWGMIMTSSICLRQTDLKSLIAYSSISHMALVVTAILIQTPWSFTGAVILMIAHGLTSSLLFCLANSNYERTHSRIMILSQGLQTLLPLMAFWWLLASLANLALPPTINLLGELSVLVTTFSWSNITLLLTGLNMLVTALYSLYMFTTTQWGSLTHHINNMKPSFTRENTLMFMHLSPILLLSLNPDIITGFSS\")\n",
    "ND4L = Seq('MPLIYMNIMLAFTISLLGMLVYRSHLMSSLLCLEGMMLSLFIMATLMTLNTHSLLANIVPIAMLVFAACEAAVGLALLVSISNTYGLDYVHNLNLLQC')\n",
    "ND5 = Seq('MTMHTTMTTLTLTSLIPPILTTLVNPNKKNSYPHYVKSIVASTFIISLFPTTMFMCLDQEVIISNWHWATTQTTQLSLSFKLDYFSMMFIPVALFVTWSIMEFSLWYMNSDPNINQFFKYLLIFLITMLILVTANNLFQLFIGWEGVGIMSFLLISWWYARADANTAAIQAILYNRIGDIGFILALAWFILHSNSWDPQQMALLNANPSLTPLLGLLLAAAGKSAQLGLHPWLPSAMEGPTPVSALLHSSTMVVAGIFLLIRFHPLAENSPLIQTLTLCLGAITTLFAAVCALTQNDIKKIVAFSTSSQLGLMMVTIGINQPHLAFLHICTHAFFKAMLFMCSGSIIHNLNNEQDIRKMGGLLKTMPLTSTSLTIGSLALAGMPFLTGFYSKDHIIETANMSYTNAWALSITLIATSLTSAYSTRMILLTLTGQPRFPTLTNINENNPTLLNPIKRLAAGSLFAGFLITNNISPASPFQTTIPLYLKLTALAVTFLGLLTALDLNYLTNKLKMKSPLCTFYFSNMLGFYPSITHRTIPYLGLLTSQNLPLLLLDLTWLEKLLPKTISQHQISTSIITSTQKGMIKLYFLSFFFPLILTLLLIT')\n",
    "ND6 = Seq('MMYALFLLSVGLVMGFVGFSSKPSPIYGGLVLIVSGVVGCVIILNFGGGYMGLMVFLIYLGGMMVVFGYTTAMAIEEYPEAWGSGVEVLVSVLVGLAMEVGLVLWVKEYDGVVVVVNFNSVGSWMIYEGEGSGLIREDPIGAGALYDYGRWLVVVTGWTLFVGVYIVIEIARGN')\n",
    "\n",
    "ref_prots = [CO1,CO2,CO3,ATP6,ATP8,CYTB,ND1,ND2,ND3,ND4,ND4L,ND5,ND6]\n",
    "ref_prots_str = ['CO1','CO2','CO3','ATP6','ATP8','CYTB','ND1','ND2','ND3','ND4','ND4L','ND5','ND6']\n",
    "\n",
    "i = 0\n",
    "for prot in ref_prots:\n",
    "    H_score = protein_aligner(prot,seq_prots_H)\n",
    "    L_score = protein_aligner(prot,seq_prots_L)\n",
    "    if H_score > L_score:\n",
    "        print('The refprot', ref_prots_str[i],'is on the Heavy strand')\n",
    "    else:\n",
    "        print('The refprot', ref_prots_str[i],'is on the Light strand')\n",
    "    \n",
    "    print('H-Strand: all ORFS and', ref_prots_str[i], '. best score is:', protein_aligner(prot,seq_prots_H))\n",
    "    print('L-Strand: all ORFS and', ref_prots_str[i], '. best score is:', protein_aligner(prot,seq_prots_L))\n",
    "    i += 1\n",
    "    \n",
    "#protein_aligner(ND4,seq_prots_H)\n",
    "#protein_aligner(ND4,seq_prots_L)\n",
    "#protein_aligner(MTND4,seq_prots_H_nouv)\n",
    "#protein_aligner(MTND4,seq_prots_L_nouv)\n",
    "#protein_aligner(MTND4,seq_prots_H_cyt)\n",
    "#protein_aligner(MTND4,seq_prots_L_cyt)\n",
    "\n",
    "#protein_aligner(CO1,seq_prots_H)\n",
    "#protein_aligner(CO1,seq_prots_L)\n",
    "#protein_aligner(MTND4,seq_prots_H_nouv)\n",
    "#protein_aligner(MTND4,seq_prots_L_nouv)\n",
    "#protein_aligner(MTND4,seq_prots_H_cyt)\n",
    "#protein_aligner(MTND4,seq_prots_L_cyt)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa6feffc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7ceeac4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
